Command tree and exported env in kubectl plugins

fabianofranz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

This is part of `kubectl` plugins V1:
- Adds support to several env vars passing context information to the plugin. Plugins can make use of them to connect to the REST API, access global flags, get the path of the plugin caller (so that `kubectl` can be invoked) and so on. Exported env vars include
  - `KUBECTL_PLUGINS_DESCRIPTOR_*`: the plugin descriptor fields
  - `KUBECTL_PLUGINS_GLOBAL_FLAG_*`: one for each global flag, useful to access namespace, context, etc
  - ~`KUBECTL_PLUGINS_REST_CLIENT_CONFIG_*`: one for most fields in `rest.Config` so that a REST client can be built.~
  - `KUBECTL_PLUGINS_CALLER`: path to `kubectl`
  - `KUBECTL_PLUGINS_CURRENT_NAMESPACE`: namespace in use
- Adds support for plugins as child of other plugins so that a tree of commands can be built (e.g. `kubectl myplugin list`, `kubectl myplugin add`, etc)

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Added support to a hierarchy of kubectl plugins (a tree of plugins as children of other plugins).

Added exported env vars to kubectl plugins so that plugin developers have access to global flags, namespace, the plugin descriptor and the full path to the caller binary.
```
@kubernetes/sig-cli-pr-reviews

fabianofranz committed:
kubectl plugins have access config, global flags and environment

fabianofranz committed:
Adds support to a tree hierarchy of kubectl plugins

fabianofranz commented:
@k8s-bot bazel test this

fabianofranz commented:
@monopole a couple pieces for V1 here.

monopole commented on pkg/kubectl/cmd/plugin.go:
```diff
@@ -61,36 +65,122 @@ func NewCmdPlugin(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cobra.Co
+		return []string{}, err
+	}
+	return []string{fmt.Sprintf("%s=%s", "KUBECTL_PLUGINS_CURRENT_NAMESPACE", cmdNamespace)}, nil
+}
```
Nice job on the env provider interface and various impls!   Very clear.

One nit: at this level it makes sense to create an array of pair struct {n, v string} and push the 
```
fmt.Sprintf("%s=%s",
```
(which occurs several times here) down into runner.Run, converting to the format exec.Command seems to need just before it is invoked. That odd detail shouldn't leak up this high.

fabianofranz replied:
I'm +1 about a struct pair, the caveat is that `os.Environ()` already returns `[]string` and it's used in one of the env providers. I'd need to iterate over it and split to make the pairs that would be returned in the provider list. Do you think it's worth?

monopole replied:
Sigh - didn't see that.  Up to you.  Seems a bit silly to unpack os.Environ.
OTHO, is osEnviron needed? I'd think a plugin running in subprocess has access to exported vars present in the parent process.

fabianofranz replied:
Setting `os.Environ` is needed, the subprocess only takes what you set there.

monopole reviewed (changes requested):
nice progress

fabianofranz commented on pkg/kubectl/cmd/plugin.go:
```diff
@@ -61,36 +65,122 @@ func NewCmdPlugin(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cobra.Co
+	env = append(env, plugins.FieldToEnv("APIPath", p.cfg.APIPath, prefix))
+	env = append(env, plugins.FieldToEnv("Prefix", p.cfg.Prefix, prefix))
+	env = append(env, plugins.FieldToEnv("Username", p.cfg.Username, prefix))
+	env = append(env, plugins.FieldToEnv("Password", p.cfg.Password, prefix))
```
I'd like to provide plugin authors a way to access the api (other that invoking `kubectl` through the `KUBECTL_PLUGINS_CALLER` env we are introducing here), so both ideas work for me. I agree with the concerns of these sensitive data passing around and the proxy is definitely easier for plugin authors, but do you think it would be easier to get approved? If so, that was part of my original plugins R&D and should not be too hard to adjust/improve be added here or in a follow-up PR.

monopole replied:
@pwittrock might weigh in.  Personally I'd rather approve a   PR that set up the proxy for the plugin and fed the plugin what it needed to know about the proxy via env vars, than the above code which is simple but enables bad behavior.  A proxy would encourage simpler plugins that followed a particular recognizable pattern.

monopole replied:
as you say -  you did a bunch of research on the proxy idea already...  just dust it off in a separate PR.

fabianofranz commented on pkg/kubectl/plugins/loader.go:
```diff
@@ -88,8 +88,15 @@ func (l *DirectoryPluginLoader) Load() (Plugins, error) {
+				setSource(path, fileInfo, child)
+			}
+		}
+		setSource(path, fileInfo, plugin)
```
I have subtrees covered in `test-cmd`, will also add some unit tests.

monopole replied:
kthnks.

fabianofranz replied:
I have them in 2 separate commits here, but I'll remember that for the next ones. ;)

fabianofranz commented:
@monopole Thanks for the review, PR updated. Addressing comments, more tests added, and the config env provider was removed. An additional way to access the API (likely through proxy) will be in a separate PR.

monopole commented:
/lgtm

Thanks!

fabianofranz commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

fabianofranz commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
